### PR TITLE
Handle promoted calls to parenless functions

### LINKED ISSUE #12736: Domain(string).length() hits compiler assertion
Special/Edge Case where calling `length()` on a `domain(string)` seems to trigger an assertion here: https://github.com/chapel-lang/chapel/blob/1426f363c00a415bc3ca77cf4033689031759f64/compiler/resolution/wrappers.cpp#L1653 

```chpl
var D: domain(string);
D.length();
```

[TIO](https://tio.run/##S85ILEjN@f@/LLFIwcVKISU/NzEzT6O4pCgzL13TmstFLyc1L70kQ0PT@v9/AA)

Note that the following does work...

```chpl
var D: domain(string);
D.length;
```

[TIO](https://tio.run/##S85ILEjN@f@/LLFIwcVKISU/NzEzT6O4pCgzL13TmstFLyc1L70kw/r/fwA)

**Found By @ajoshpratt**

### PR DESCRIPTION
**Pull request by vasslitvinov**
Resolves #12736.

With this change, promoted calls to parenless functions
are handled properly.

While this corner case could be handled more easily by replacing
the INT_ASSERT (that was tripped over) with an immediate return
from addSetIteratorShape(), the current code has the advantage
of preserving the shape of the promotion leader.

Other implementation specifics are described in comments.

### CODE REVIEW

File: compiler/resolution/wrappers.cpp [hunk 1]
```diff
@@ -2017,14 +2017,70 @@ static Symbol* leadingArg(PromotionInfo& promotion, CallExpr* call) {
...
+  // Otherwise, this is a nested call, see the comments above.
+  isNestedCall = true;
+  move = parent2;
+  return false;
+}
+
+// Returns true if we should not be setting this IR's shape.
+// See the comment on dontSetIteratorShape1().
+static bool dontSetIteratorShape2(Symbol* irTemp, bool isNestedCall) {
+  static std::set<Symbol*> outIRtemps;
```
Comment (e-kayrakli) [inline compiler/resolution/wrappers.cpp:2060]: Do we have static local variables in the compiler? I wouldn't oppose having them, but I wouldn't feel brave enough to add them myself if they don't exist :)

File: compiler/resolution/wrappers.cpp [hunk 2]
```diff
@@ -2017,14 +2017,70 @@ static Symbol* leadingArg(PromotionInfo& promotion, CallExpr* call) {
...
+// for the outer call and we do not yet have the IR for the inner call.
+// So we cannot "set shape" on this inner IR. This is most likely OK
+// because the inner IR feeds directly into the outer call. The outer call
+// should also be promoted and inherit the shape from the inner call.
+//
+// We will go ahead and set the shape for 'irTemp' right away from the shape
+// of the inner call. We will also remember 'irTemp' to avoid setting its
+// shape the second time, which would not work. See dontSetIteratorShape2().
+//
+static bool dontSetIteratorShape1(CallExpr* &move, CallExpr* call,
```
Comment (e-kayrakli) [inline compiler/resolution/wrappers.cpp:2035]: Can we rename this to `adjustMoveForSetIteratorShape` as I believe that's the most important thing it is doing. Then the fact that it returns bool can be confusing, if you think that way, you can have an argument `bool &settable` or something.

**Review by e-kayrakli — APPROVED** [review]
I would probably implement this with two additional functions that does:

- one that adjusts the `move` to look at,
- one that checks whether we can add `set iterator shape`

Currently the `dontSetIteratorShape1` seem to be doing a little bit of both.

But I may be missing something, and would be fine with this 2-phase-ish approach. Could you adjust the function names as in the inline comments?

Looks good otherwise!

File: compiler/resolution/wrappers.cpp (hunk 1 above)
Comment (vasslitvinov) [reply compiler/resolution/wrappers.cpp:2060]: There are a couple of places where we do this already. No innovation here, alas. :)

**Review by e-kayrakli — APPROVED** [review]
Looks great!

**Pull request merged**